### ISSUE TITLE
Is it possible to use css rewrite in the first pageload?

### ISSUE THREAD

**Issue by onnimonni** [pagespeed/ngx_pagespeed#1306]
**TL;DR** Is it possible to rewrite `*.css` files immediately in the first pageload?

It doesn't matter if it's little bit slower and blocking. It's just one pageload.

My use case is that in WordPress themes need to include `style.css` which has certain mandatory comments in the beginning of the theme. Here's an example:
```
/*
Theme Name: Twenty Fifteen
Theme URI: https://wordpress.org/themes/twentyfifteen/
Author: the WordPress team
Author URI: https://wordpress.org/
Description: Our 2015 default theme is clean, blog-focused, and designed for clarity. Twenty Fifteen's simple, straightforward typography is readable on a wide variety of screen sizes, and suitable for multiple languages. We designed it using a mobile-first approach, meaning your content takes center-stage, regardless of whether your visitors arrive by smartphone, tablet, laptop, or desktop computer.
Version: 1.6
License: GNU General Public License v2 or later
License URI: http://www.gnu.org/licenses/gpl-2.0.html
*/
```

If this block is not included the theme will not work properly.

It includes the version number which we don't want to expose and sometimes developers might put something sensitive over there accidentally and in case that happens we don't want to reveal that to the attacker.

ngx_pagespeed is the best option I have found to consistently strip comments away. My only problem is that for the first pageload nginx doesn't alter any stylesheets.

I understand that some features need the beacon to be in place but I'm mainly considering the security enhancement that `pagespeed EnableFilters rewrite_css;` adds to our use case.

**Comment by Lofesa** [comment]
Perhaps trying this

[https://developers.google.com/speed/pagespeed/module/system#rewrite_deadline](https://developers.google.com/speed/pagespeed/module/system#rewrite_deadline)

**Comment by jeffkaufman** [comment]
Guaranteed rewriting the first load and every load isn't something PageSpeed was designed for, so any solutions here aren't going to be ideal.

I think the most reliable way to do this is to use [LoadFromFile](https://developers.google.com/speed/pagespeed/module/domains#ModPagespeedLoadFromFile) with:

```
pagespeed InPlaceWaitForOptimized on;
pagespeed InPlaceRewriteDeadlineMs [something big];
```

The problem is, this isn't good enough for security because people can get around it.  For example, they can request `example.css?PageSpeed=off` and get the original, even if you have set a [RequestOptionOverride](https://developers.google.com/speed/pagespeed/module/experiment#restrict-request-options) token.  And they can create a url like `example.css.pagespeed.ce.0.css` [1] even if you have [Forbidden](https://developers.google.com/speed/pagespeed/module/config_filters#enabling) the use of `extend_cache`.

[1] It doesn't have to be `ce`; other filters like `jm` or `cc` will work there too.  It's a consequence of how PageSpeed manages its fallback path when it gets a request for an optimized request it can't serve from cache.

**Comment by onnimonni** [comment]
I know that this isn't going to protect us against persistant attacker. Persistant attacker can also deduce the original theme version by just looping all themes by using this configuration to collect the matching the checksums but that takes plenty of more resources for the attacker.

The idea behind this is to make attack surface of automated attacks harder.

Is it okay if I fix the things that you mentioned here by adding these?

```
if ($arg_pagespeed = "off") { deny all; }
if ($arg_pagespeedspeedfilters != "") { deny all; }
rewrite ^(.*)style.css.pagespeed(.*)$ $1/style.css;
```

**Comment by onnimonni** [comment]
I think we will just deny `style.css` and move the required styles into another file. Thanks for your help :).

**Comment by jeffkaufman** [comment]
In case someone comes back to it later, there are some problems with these:

    if ($arg_pagespeed = "off") { deny all; }
    if ($arg_pagespeedspeedfilters != "") { deny all; }

These two are fine.

    rewrite ^(.*)style.css.pagespeed(.*)$ $1/style.css;

By default PageSpeed will use `rewrite_css` to create resources like `styles/A.yellow.css.pagespeed.cf.KM5K8SbHQL.css` in order to extend the cache lifetime of resources.  So if you're going to do that kind of rewriting, and are ok losing the longcaching benefits, it would be better to pair it with [CssPreserveUrls](https://developers.google.com/speed/pagespeed/module/config_filters#preserveurls) to make sure PageSpeed doesn't generate them in the first place.

**Comment by onnimonni** [comment]
Thanks for the follow up. I tried this and in my testing it actually works really well.

Here's minimal configuration that you need for WordPress:
```
# Use pagespeed assets straight from corresponding files
pagespeed LoadFromFileMatch "^https?://[^/]+/wp-content/themes/" "/var/www/mysite/wp-content/themes/";

# Always rewrite style.css to get rid of the mandatory comments for WordPress
location ~ ^.*/themes/[^/]+/style\.css {
    
    ## Deny overriding Pagespeed
    if ($arg_pagespeed != '') { return 403; }
    if ($arg_pagespeedspeedfilters != '') { return 403; }

    # Don't allow overriding rewrite_css filter with pagespeed fixed urls like style.css.pagespeed.ce.0.css
    # Source: https://github.com/pagespeed/ngx_pagespeed/issues/1306
    location ~ ^(.*)/themes/([^/]+)/style.*pagespeed.* {
        return 301 $1/themes/$2/style.css;
    }

    # Preserve this url
    pagespeed CssPreserveUrls true;

    # Minify and remove comments from css
    # This is used to get rid of the comments in the beginning of the file
    pagespeed EnableFilters rewrite_css;

    # Wait that nginx has prepared the file
    pagespeed InPlaceWaitForOptimized on;

    # Don't ever try to load non modified original style.css
    pagespeed InPlaceRewriteDeadlineMs 100000000;

}
```

@jeffkaufman How bad is that `LoadFromFileMatch` performance wise?

Is there a configuration where I can use LoadFromFile which matches multiple hostnames or $http_host?

**Comment by jeffkaufman** [comment]
Both the `LoadFromFileMatch` and `ProcessScriptVariables` should have negligible performance impact; they're both compiled (regexp and nginx variable script) and execute very quickly.
